beetsplug/bandcamp/__init__.py

```diff
@@ -216,7 +216,21 @@ def candidates(
         if "various" in artist.lower():
             artist = ""
 
-        search = {"query": album, "artist": artist, "label": label, "search_type": "a"}
+        if album == "":
+            search = {
+                "query": item.title,
+                "artist": artist,
+                "label": label,
```

Comment:
Could you give an example `album` and `title` that you tested this with?

This is going to return a track URL and I wonder whether `self.get_album_info` handles it well returning the actual album instead of a single track?

Reply:
For example ```999999999 - RAVE 4 LOVE``` yields search results ranking the correct result on first place (```RAVE REWORKS```). Then you still have to pass the correct number because identifying the album from a single combination of artist and title wont get incredibly high confidence levels but its enough to make the search a lot less tedious. Here is what the verbose output looks like:

```log
(beetcamp-py3.12) philip@MacBook-Pro-von-Philip beetcamp % beet -v import '/[...]/999999999 - RAVE 4 LOVE.m4a'
user configuration: /Users/philip/.config/beets/config.yaml
data directory: /Users/philip/.config/beets
plugin paths:
Sending event: pluginload
library database: /Users/philip/Music/Library/Beets/musiclibrary.db
library directory: /Users/philip/Music/Library/Music
Sending event: library_opened
Sending event: import_begin
Sending event: import_task_created
Sending event: import_task_start
Looking up: /[...]/999999999 - RAVE 4 LOVE.m4a
Tagging 999999999  -
No album ID found.
Search terms: 999999999  -
Album might be VA: False
bandcamp: Searching releases of type 't' for query ' RAVE 4 LOVE' using '{'query': ' RAVE 4 LOVE', 'artist': '999999999 ', 'label': '', 'search_type': 't'}'
Sending event: albuminfo_received
Candidate: 999999999 - RAVE REWORKS (https://999999999music.bandcamp.com/album/rave-reworks)
Computing track assignment...
...done.
Success. Distance: 0.48
Sending event: albuminfo_received
Candidate: 999999999 - RAVE REWORKS (https://999999999music.bandcamp.com/album/rave-reworks#p1901314029)
Computing track assignment...
...done.
Success. Distance: 0.48
Sending event: albuminfo_received
Candidate: 999999999 - RAVE REWORKS (https://999999999music.bandcamp.com/album/rave-reworks)
Duplicate.
Sending event: albuminfo_received
Candidate: 999999999 - RAVE REWORKS (https://999999999music.bandcamp.com/album/rave-reworks#p1901314029)
Duplicate.
Sending event: albuminfo_received
Candidate: Various Artists - Wonderland (https://absolute.bandcamp.com/album/wonderland)
Computing track assignment...
...done.
Success. Distance: 0.96
Sending event: albuminfo_received
Candidate: Various Artists - Wonderland (https://absolute.bandcamp.com/album/wonderland#p4289721344)
Computing track assignment...
...done.
Success. Distance: 0.96
Sending event: albuminfo_received
Candidate: Huerco S. - For Those Of You Who Have Never (And Also Those Who Have) (https://brianleeds.bandcamp.com/album/for-those-of-you-who-have-never-and-also-those-who-have)
Computing track assignment...
...done.
Success. Distance: 0.97
Sending event: albuminfo_received
Candidate: Move D - House Grooves, Vol. 1 (https://s4aw.bandcamp.com/album/house-grooves-vol-1-remastered)
Computing track assignment...
...done.
Success. Distance: 0.92
Sending event: albuminfo_received
Candidate: Move D - House Grooves, Vol. 1 (https://s4aw.bandcamp.com/album/house-grooves-vol-1-remastered#p4244096095)
Computing track assignment...
...done.
Success. Distance: 0.92
Evaluating 7 candidates.

/[...]/999999999 - RAVE 4 LOVE.m4a (1 items)
Sending event: import_task_before_choice
Sending event: before_choose_candidate
Finding tags for album "999999999  - ".
Candidates:
1. 999999999 - RAVE REWORKS (52.4%) (album, missing tracks, tracks) (bandcamp, Digital Media, 2018, IT, NineTimesNine, NTNLTD002)
2. 999999999 - RAVE REWORKS (52.4%) (album, missing tracks, tracks) (bandcamp, Vinyl, 2018, IT, NineTimesNine, NTNLTD002)
3. Move D - House Grooves, Vol. 1 (8.1%) (album, artist, missing tracks, ...) (bandcamp, Digital Media, 2021, DE, 		All That Jelly, ATJ 008)
4. Move D - House Grooves, Vol. 1 (8.1%) (album, artist, missing tracks, ...) (bandcamp, Vinyl, 2021, DE, 		All That Jelly, ATJ 008)
5. Various Artists - Wonderland (3.7%) (missing tracks, album, tracks) (bandcamp, Digital Media, 2021, GB, ABSOLUTE.)
6. Various Artists - Wonderland (3.7%) (missing tracks, album, tracks) (bandcamp, Cassette, 2021, GB, ABSOLUTE.)
7. Huerco S. - For Those Of You Who Have Never (And Also Those Who Have) (3.3%) (missing tracks, album, artist, ...) (bandcamp, Digital Media, 2016, US, Brian Leeds)
# selection (default 1), Skip, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort? 1
```

Reply:
Nice, it seems like `beetcamp` handles this well and still returns the album.

What do other autotagger return in comparison, for example MusicBrainz, discogs? While this makes life a bit easier if `album` is not set in music files being imported, I am slightly worried about this behaviour being inconsistent with how other autotaggers work.

As a user, I would probably expect all autotaggers return rubbish when album is missing, instead of seeing _only_ bandcamp returning something sensible.

Reply:
I tested using the search feature with the follwing inputs: ```artist=999999999``` and ```album=LOVE 4 RAVE```. MusicBrainz seems to handle it well even though confidence drops quite severly:
```log
Searching for MusicBrainz releases with: {'release': 'love 4 rave', 'artist': '999999999', 'tracks': '1'}
[...]
Finding tags for album " - ".
  Candidates:
  1. (10.4%) Various Artists - Wave 4 Rave
             ≠ album, missing tracks, tracks
             MusicBrainz, Digital Media, 2020, IT, NineTimesNine, NTNVA001, None
  2. (9.0%) 999999999 - Rave Reworks
             ≠ album, artist, tracks, ...
             MusicBrainz, 12" Vinyl, 2018, IT, NineTimesNine, NTNLTD002, None
  3. (9.0%) 999999999 - Rave Reworks
             ≠ album, artist, tracks, ...
             MusicBrainz, Digital Media, 2018, XW, None, None, None
  4. (8.2%) 999999999 - 000000002
             ≠ album, artist, missing tracks, ...
             MusicBrainz, Digital Media, 2016, XW, None, None, None
  5. (0.0%) 999999999 - 2023-05-20: BBC Radio 1 Essential Mix
             ≠ album, artist, tracks
             MusicBrainz, Digital Media, 2023, GB, None, None, None
```
Same thing with Discogs:
```log
Finding tags for album " - ".
  Candidates:
  1. (12.8%) 999999999 - Rave Reworks
             ≠ album, artist, tracks, ...
             Discogs, Vinyl, 2018, Italy, NineTimesNine, NTNLTD002, None
  2. (2.9%) Viktor Gambon - Dark Techno Sessions Part 08
             ≠ missing tracks, album, artist, ...
             Discogs, File, 2020, Netherlands, Dark Techno Sessions, DTS008, None
```
Spotify also works but returns different metadata as this track is released as individual single there:
```log
spotify: Searching Spotify for 'album:LOVE 4 RAVE artist:999999999'
[...]
Finding tags for album " - ".
  Candidates:
  1. (18.6%) 999999999 - LOVE 4 RAVE
             ≠ album, artist, tracks, ...
             Spotify, None, 2019, None, NineTimesNine, None, None
  2. (12.9%) 999999999 - Rave 4 love
             ≠ album, artist, tracks, ...
             Spotify, None, 2020, None, 999999999, None, None
```

Reply:
I see. As far as I uunderstand your change here handles import of the file that *does not have **album** set*, right?

Given your test above

> I tested using the search feature with the follwing inputs: artist=999999999 and album=LOVE 4 RAVE.

I'd expect `beetcamp` to return the correct album even without the change that you added. Could you confirm this?

Reply:
No it doesn't return the correct album because the search function tells beetcamp to explicitly look for albums where `title=query` but when passing my modified query `bandcamp: Searching releases of type 't' for query 'title' using '{'query': 'title', 'artist': 'artist', 'label': '', 'search_type': 't'}'` because of the search type results are in general more relevant. For example when using search this is the ranking for `artist=999999999` and `album=LOVE 4 RAVE`:

```log
Search terms: 999999999 - LOVE 4 RAVE
bandcamp: Searching releases of type 'a' for query 'LOVE 4 RAVE' using '{'query': 'LOVE 4 RAVE', 'artist': '999999999', 'label': '', 'search_type': 'a'}'
Candidates:
  1. (15.5%) Dave Linnenbank - Love Songs for Strangers
             ≠ album, missing tracks, artist, ...
             bandcamp, Digital Media, 2023, DE, Dave Linnenbank, , None
  2. (8.7%) Ace Warrior - Burning For Your Love EP
             ≠ album, artist, missing tracks, ...
             bandcamp, Digital Media, 2020, IT, Dave Rodgers Music, , None
  3. (8.7%) Ace Warrior - Burning For Your Love EP
             ≠ album, artist, missing tracks, ...
             bandcamp, Digital Media, 2020, IT, Dave Rodgers Music, , None
  4. (7.0%) Amanda Duncan - Love I Have for You
             ≠ missing tracks, album, artist, ...
             bandcamp, Digital Media, 2009, US, Amanda Duncan, , None
  5. (3.7%) Posthuman - Requiem For A Rave
             ≠ missing tracks, album, artist, ...
             bandcamp, Digital Media, 2021, GB, I Love Acid / Balkan, BV50T, None
  6. (3.7%) Posthuman - Requiem For A Rave
             ≠ missing tracks, album, artist, ...
             bandcamp, 2xVinyl, 2021, GB, I Love Acid / Balkan, BV50, None
  7. (3.7%) Posthuman - Requiem For A Rave
             ≠ missing tracks, album, artist, ...
             bandcamp, Cassette, 2021, GB, I Love Acid / Balkan, BV50T, None
  8. (2.9%) Miss Tess & The Talkbacks - The Love I Have For You
             ≠ missing tracks, album, artist, ...
             bandcamp, Digital Media, 2013, US, Signature Sounds Recording Inc., , None
  9. (2.9%) Miss Tess & The Talkbacks - The Love I Have For You
             ≠ missing tracks, album, artist, ...
             bandcamp, CD, 2013, US, Signature Sounds Recording Inc., , None
```

Reply:
Of course this use case might be niche so I would understand if this is not feasible as default behavior. My suggestion would be to include a config setting which is disabled by default. I'd be happy to write the implementation for that!